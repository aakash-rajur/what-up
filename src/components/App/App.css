@import url("../../utils/variables.css");

.accent-font-color {
    color: var(--accent-color);
}

#root {
    text-align: center;
    color: gray;
    overflow: auto;
    display: flex;
    flex-direction: column;
}

#root > * {
    text-align: initial;
}

#root .title {
    font-weight: 100;
    font-size: 6em;
    text-align: center;
    margin: 0.25em 0;
    transition: all 0.3s;
}

.new-task-container {
    display: grid;
    grid-template: auto/auto;
    grid-gap: 8px;
    justify-content: center;
    position: sticky;
    top: 0;
    width: 100%;
    z-index: 1;
}

.new-task {
    border: none;
    padding: 0.5em 1em 0.5em 2em;
    font-size: 2em;
    font-weight: 100;
    border-radius: 4px;
    color: gray;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
    background: white;
    transition: all 0.3s cubic-bezier(.25, .8, .25, 1);
    grid-row: 1;
    grid-column: 1;
}

.new-task:hover, .new-task:focus {
    box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
}

.check-all {
    grid-row: 1;
    grid-column: 1;
    align-self: center;
    margin-left: 1.5em;
    opacity: 0.5;
}

.filter-container {
    display: grid;
    grid-template: auto/repeat(4, auto);
    grid-gap: 8px;
    justify-content: center;
}

#root .documentation {
    text-align: center;
    font-size: 0.7em;
    padding: 2em 0;
    line-height: 2em;
    font-weight: 100;
    color: lightgray;
}

#root .documentation em {
    font-weight: 100;
}

#root .documentation a {
    color: lightblue;
}

@media (max-width: 600px) {
    #root .title {
        font-size: 4em;
    }

    .new-task-container {
        grid-template: auto/1fr;
    }

    .new-task {
        font-size: 1.5em;
    }

    .check-all {
        margin-left: 2.5em;
    }

    .task-list {
        grid-template: auto/1fr;
    }
}

@media (pointer: coarse) and (min-width: 450px) {
    .filter-container {
    }
}



