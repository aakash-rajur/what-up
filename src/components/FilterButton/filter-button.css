@import "../../utils/variables.css";

.filter {
    background: none;
    border: none;
    font-size: 0.8em;
    font-weight: 100;
    padding: 0.5em 2em 0.5em 1em;
    border-radius: 4px;
    cursor: pointer;
    color: white;
    transition: all 0.3s;
    outline: none;
    background: lightgray;
    box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
}

.filter:hover {
    box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
    background: gray;
}

.filter[data-active="true"] {
    box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
    background: deepskyblue;
    color: white;
    cursor: not-allowed;
}


.badge {
    position: absolute;
    background: white;
    color: var(--accent-opaque);
    width: 16px;
    height: 16px;
    border-radius: 50%;
    right: 4px;
    top: calc(50% - 8px);
    font-size: 0.8em;
    display: flex;
    align-items: center;
    justify-content: center;
}

@media (pointer: coarse) {
    .filter {
        font-size: 1em;
    }
}
