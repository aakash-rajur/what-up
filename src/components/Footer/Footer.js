import React from 'react';

function Footer() {
	return (
		<div className="documentation">
			<em>double-click to edit tasks</em><br/>
			Created by <a href="https://github.com/aakashRajur"
			              className="link">aakashRajur</a><br/>
			Inspired by <a href="http://todomvc.com/examples/react/" className="link">React â€¢ TodoMVC</a><br/>
		</div>
	);
}

export default Footer;