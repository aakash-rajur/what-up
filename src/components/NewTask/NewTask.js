import PropTypes from 'prop-types'
import React from 'react';
import {withNewTaskAddition} from "../../utils/apollo";
import Input from "../Input/Input";

export function NewTask({newTask, onChange, addNewTask, ...rest}) {
	let inputProps = {...rest};
	delete inputProps.onDone;
	return (
		<Input value={newTask} onChange={onChange}
		       onValueReturn={addNewTask} {...inputProps}/>
	);
}

NewTask.propTypes = {
	newTask: PropTypes.string,
	onChange: PropTypes.func
};

export default withNewTaskAddition(NewTask);
