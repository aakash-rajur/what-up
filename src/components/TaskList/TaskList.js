import PropTypes from 'prop-types';
import React, {Component} from 'react';
import {withQueryTasks} from "../../utils/apollo";
import {TASK_ALL} from "../../utils/constants";
import Error from "../Error/Error";
import Loading from "../Loading/Loading";
import Task from "../Task/Task";
import './task-list.css';

export class TaskList extends Component {
	static propTypes = {
		tasks: PropTypes.array,
		loading: PropTypes.bool,
		error: PropTypes.any,
		filter: PropTypes.string
	};

	state = {
		tasks: [],
		loading: true
	};

	componentDidUpdate(prevProps) {
		if (this.props.loading !== prevProps.loading) {
			this.setState({loading: this.props.loading && !this.state.tasks.length});
		}
		if (this.props.tasks !== prevProps.tasks && this.props.tasks)
			this.setState({tasks: this.props.tasks, loading: false});
	}

	render() {
		let {error, filter} = this.props, {
			loading, tasks
		} = this.state;
		return (
			error ? <Error text={error && (typeof error === "string" ? error : JSON.stringify(error))}/> :
				loading ? (<div className="loading-container"><Loading/></div>) :
					!tasks.length ? (
						<div className="no-tasks">
							{filter === TASK_ALL ? 'all your tasks show up here!' :
								`your ${filter.toLowerCase()} tasks show up here!`}
						</div>
					) : (
						<div className="task-list">
							{tasks.map((task, index) => <Task {...task} index={index} key={index}/>)}
						</div>
					)
		)
	}
}

export default withQueryTasks(TaskList);
