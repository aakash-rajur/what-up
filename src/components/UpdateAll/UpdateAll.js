import PropTypes from "prop-types";
import React from "react";
import CHEVRON_DOWN_ICON from "../../assets/chevron-down.svg";
import {withUpdateAll} from "../../graphql/hoc";

export function UpdateAll({className, updateAll}) {
  return (
    <img
      src={CHEVRON_DOWN_ICON}
      alt="toggle-all"
      className={className}
      onClick={updateAll}
      title="Update All"
    />
  );
}

UpdateAll.propTypes = {
  className: PropTypes.string,
  nextStatus: PropTypes.string,
  filter: PropTypes.string
};

export default withUpdateAll(UpdateAll);
