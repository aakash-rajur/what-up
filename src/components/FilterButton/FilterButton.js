import React, {Component} from 'react';
import PropTypes from 'prop-types';
import './filter-button.css'

class FilterButton extends Component {
    static propTypes = {
        filter: PropTypes.string,
        onClick: PropTypes.func,
        className: PropTypes.string,
        title: PropTypes.string,
        active: PropTypes.bool,
        stat: PropTypes.number,
    };

    static defaultProps = {
        active: false
    };

    constructor(props) {
        super(props);
        this.onClick = this.onClick.bind(this);
    }

    render() {
        let {
            title, active,
            filter, stat
        } = this.props;
        return (
            <button title={title} className={`filter ${(filter || '').toLowerCase()}`}
                    onClick={this.onClick} data-active={active} disabled={active}>
                {title}
                {stat && <span className="badge">{stat > 99 ? 'âˆž' : stat}</span>}
            </button>
        )
    }

    onClick() {
        let {onClick, filter} = this.props;
        onClick && onClick(filter);
    }
}

export default FilterButton;
