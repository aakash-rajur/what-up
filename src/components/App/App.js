import PropTypes from 'prop-types';
import React, {Component, Fragment} from 'react';
import {defaultTaskUpdated, withNotificationAndTaskSubscription} from "../../utils/apollo";
import {TASK_ALL, TASK_CANCELLED, TASK_COMPLETED, TASK_CREATED} from "../../utils/constants";
import {FILTER_BUTTON_TEMPLATE} from "../../utils/library";
import FilterButton from "../FilterButton/FilterButton";
import Footer from "../Footer/Footer";
import NewTask from "../NewTask/NewTask";
import SnackBar from "../SnackBar/SnackBar";
import TaskList from "../TaskList/TaskList";
import UpdateAll from "../UpdateAll/UpdateAll";
import './App.css';

export class App extends Component {
	static propTypes = {
		timestamp: PropTypes.string,
		stats: PropTypes.shape({
			[TASK_ALL]: PropTypes.number,
			[TASK_CANCELLED]: PropTypes.number,
			[TASK_COMPLETED]: PropTypes.number,
			[TASK_CREATED]: PropTypes.number
		}),
		notification: PropTypes.shape({
			action: PropTypes.string,
			data: PropTypes.any
		})
	};
	
	static defaultProps = {
		timestamp: '0',
		stats: {...defaultTaskUpdated},
		notification: {action: 'DEFAULT', data: null}
	};
	
	constructor(props) {
		super(props);
		console.log(props);
		this.onFilterChange = this.onFilterChange.bind(this);
		this.onNewTaskChange = this.onNewTaskChange.bind(this);
		this.clearNewTaskInput = this.clearNewTaskInput.bind(this);
		this.getNextUpdateALlStatus = this.getNextUpdateALlStatus.bind(this);
		this.state = {
			filter: TASK_ALL,
			newTask: '',
			message: ''
		};
	}
	
	async componentDidMount() {
		window.addEventListener('unload', () =>
			document.cookie = `connection=;expires=Thu, 01 Jan 1970 00:00:01 GMT;`);
		console.log(await (await fetch('http://what-up.herokuapp.com/', {
			credentials: 'include',
			mode: 'cors'
		})).text());
	}
	
	componentDidUpdate(prevProps) {
		const {
			notification: prevNotification
		} = prevProps, {
			notification
		} = this.props;
		if (prevNotification !== notification
			&& prevNotification.action !== notification.action
			&& prevNotification.timestamp !== notification.timestamp) {
			const {action, data} = notification;
			if (action === 'NEW_SESSION') {
				const {token} = data;
				document.cookie = `session=${token};`;
				document.cookie = `connection=true;`;
			} else if (action === 'SESSION_EXPIRED') {
				document.cookie = `session=;expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
			} else if (action === 'SESSION_RESTORED') {
				document.cookie = 'connection=true;'
			}
			this.setState({message: data.message});
		}
	}
	
	render() {
		let {
			filter,
			newTask,
			message
		} = this.state, {
			timestamp = 'NONE',
			stats
		} = this.props;
		return (
			<Fragment>
				<h1 className="accent-font-color title">What Up!</h1>
				<span className="new-task-container">
					<NewTask newTask={newTask} type="text"
					         autoFocus={true} placeholder="new task?"
					         className="app-width new-task"
					         onChange={this.onNewTaskChange}
					         onDone={this.clearNewTaskInput}/>
					<UpdateAll className="icon check-all"
					           nextStatus={this.getNextUpdateALlStatus()}
					           filter={filter}/>
					<div className="filter-container">
						{FILTER_BUTTON_TEMPLATE.map((type, index) => (
							<FilterButton key={index} {...type}
							              stat={stats ? stats[type.filter] : 0}
							              active={type.filter === filter}
							              onClick={this.onFilterChange}/>
						))}
	                </div>
	        </span>
				<TaskList filter={filter} timestamp={timestamp}/>
				<Footer/>
				<SnackBar
					timeout={2000}
					message={message}
				/>
			</Fragment>
		);
	}
	
	onNewTaskChange({target: {value}}) {
		this.setState({newTask: value});
	}
	
	onFilterChange(filter) {
		this.setState({filter});
	}
	
	clearNewTaskInput() {
		this.setState({newTask: ''});
	}
	
	getNextUpdateALlStatus() {
		if (!this.props || !this.props.stats)
			return TASK_CREATED;
		const {
			[TASK_ALL]: total,
			[TASK_COMPLETED]: completed
		} = this.props.stats, {
			filter
		} = this.state;
		if (filter === TASK_ALL)
			return completed === total ? TASK_CREATED : TASK_COMPLETED;
		switch (filter) {
			case TASK_ALL:
				return completed === total ? TASK_CREATED : TASK_COMPLETED;
			case TASK_CREATED:
				return TASK_COMPLETED;
			case TASK_COMPLETED:
			case TASK_CANCELLED:
			default:
				return TASK_CREATED;
		}
	}
}

export default withNotificationAndTaskSubscription(App);
