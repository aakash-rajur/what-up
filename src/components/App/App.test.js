import React from "react";
import FilterButton from "../FilterButton/FilterButton";
import Footer from "../Footer/Footer";
import NewTask from "../NewTask/NewTask";
import TaskList from "../TaskList/TaskList";
import UpdateAll from "../UpdateAll/UpdateAll";
import {App} from "./App";

describe("App component", () => {
  let component = shallow(<App />);

  it("dom integrity", () => {
    [
      "h1.accent-font-color.title",
      "span.new-task-container",
      NewTask,
      UpdateAll,
      "div.filter-container",
      TaskList,
      Footer
    ].forEach(child => expect(component.find(child).exists()).toBe(true));
    expect(component.find(FilterButton).length).toBe(4);
  });

  it("match snapshot with non-empty list", () => {
    expect(component).toMatchSnapshot();
  });
});
