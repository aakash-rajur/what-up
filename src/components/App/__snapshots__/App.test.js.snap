// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component match snapshot with non-empty list 1`] = `
<Fragment>
  <h1
    className="accent-font-color title"
  >
    What Up!
  </h1>
  <span
    className="new-task-container"
  >
    <Apollo(NewTask)
      autoFocus={true}
      className="app-width new-task"
      newTask=""
      onChange={[Function]}
      onDone={[Function]}
      placeholder="new task?"
      type="text"
    />
    <Apollo(UpdateAll)
      className="icon check-all"
      filter="ALL"
      nextStatus="CREATED"
    />
    <div
      className="filter-container"
    >
      <FilterButton
        active={true}
        filter="ALL"
        key="0"
        onClick={[Function]}
        stat={0}
        text="All"
        title="View All Tasks"
      />
      <FilterButton
        active={false}
        filter="CREATED"
        key="1"
        onClick={[Function]}
        stat={0}
        text="Active"
        title="View Active Tasks"
      />
      <FilterButton
        active={false}
        filter="COMPLETED"
        key="2"
        onClick={[Function]}
        stat={0}
        text="Completed"
        title="View Completed Tasks"
      />
      <FilterButton
        active={false}
        filter="CANCELLED"
        key="3"
        onClick={[Function]}
        stat={0}
        text="Cancelled"
        title="View Cancelled Tasks"
      />
    </div>
  </span>
  <Apollo(TaskList)
    action={null}
    filter="ALL"
    timestamp="0"
  />
  <Footer />
  <SnackBar
    message={null}
    timeout={2000}
  />
</Fragment>
`;
